[SPARK-663]Implement Fair Scheduler in Spark Cluster Scheduler

xiajunluan opened the pull request:
1. What features I have implemented?
   Based on Harold's code, I mainly finished below functions
   a) support FIFO (between TaskSetManagers)or fair scheduler(between pools) in cluster scheduler,default is FIFO
   b) support FIFO or fair scheduler between TaskSetManagers within pool, default is FIFO
   c) support priority and stageId comparator for FIFO
   d) support miniShares, weight , numRunningTask comparator for fair scheduler
   e) support  to set which pool job will belong to by user when using Fair Scheduler Algorithm, default pool name is "default" if user does not set 
   f) support XML configuration for pool creation
2. How will user write spark application using fair scheduler
   spark application will looks like:
   //===============================
   /\* Set FIFO or FS mode(“FAIR” or “FIFO”, default is "FIFO") 
   */
   System.setProperty("spark.cluster.schedulingmode", “FAIR”)
   
   /_set XML configuration file path_/
   System.setProperty("spark.fairscheduler.allocation.file", path/to/XML configuration file)
   
   spark = new SparkContext("...")
   
   /_Set pool on each spark thread_/
   spark.addLocalProperties("spark.scheduler.cluster.fair.pool", poolName)
   ..........
   //===============================
3. Format of XML configuration
   <allocations>
   <pool name="pool1">
   <minShares>2</minShares>
   <weight>1</weight>
   <schedulingMode>FIFO</schedulingMode>
   </pool>
   <pool name="pool2">
   <minShares>10</minShares>
   <weight>2</weight>
   <schedulingMode>FAIR</schedulingMode>
   </pool>
   </allocations>
4. Class relationship diagram
   ![Fair Scheduler another version](https://f.cloud.github.com/assets/1168928/263707/6bd2669e-8d78-11e2-807d-09ff3d3135e1.jpg)

rold50 committed:
Made changes to the SparkContext to have a DynamicVariable for setting local properties that can be passed down the stack. Added an implementation of the fair scheduler

rold50 committed:
Changed the name of the system property to set the allocation xml

rold50 committed:
Updated/Refactored the Fair Task Scheduler. It does not inherit ClusterScheduler anymore. Rather, ClusterScheduler internally uses TaskSetQueuesManager that handles the scheduling of taskset queues. This is the class that should be extended to support other scheduling policies

rold50 committed:
Cleaned up the code

rold50 committed:
Removed some commented code

xiajunluan committed:
refactor fair scheduler implementation
1.Chage "pool" properties to be the memeber of ActiveJob
2.Abstract the Schedulable of Pool and TaskSetManager
3.Abstract the FIFO and FS comparator algorithm
4.Miscellaneous changing of class define and construction

mateiz commented:
Thanks for the detailed overview, Andrew! Will take a look at it this weekend.

xiajunluan committed:
Just update typo and comments

xiajunluan commented:
Hi Mateiz
    I would like to know if there has any comments on implementation of fair scheduler.

mateiz commented:
Hey Andrew,

The general approach looks quite good, but I have a comment on the code style -- would you mind changing it to match the style of the rest of the project? Here are our guidelines: http://spark-project.org/docs/latest/contributing-to-spark.html. Basically, we use the standard Scala style (linked there) with a few small changes. You have a lot more spacing around things than is common there.

One other small thing: in Scala, instead of getter methods (e.g. Schedulable.getWeight, etc), properties usually just have a name (e.g. def weight: Double). Subclasses can override this using either a method or a val or var, since all calls to fields are automatically method calls. So it would be good to change those in the new classes. In Spark the only places we have "get" are where we're only going to call that method once and we ask the user to compute something there (there aren't many such places).

xiajunluan commented:
Thanks, Mateiz. I will follow the guidelines and update the code style.

xiajunluan committed:
1.remove redundant spacing in source code
2.replace get/set functions with val and var defination

xiajunluan committed:
change some typo and some spacing

xiajunluan commented:
@mateiz I have updated code based on your suggestion and help to review codes again.

xiajunluan committed:
add properties default value null to fix sbt/sbt test errors

AmplabJenkins commented:
Can one of the admins verify this patch?

mateiz commented on core/src/main/scala/spark/scheduler/cluster/FairTaskSetQueuesManager.scala:
```diff
@@ -0,0 +1,157 @@
+  val POOL_POOLS_PROPERTY = "pool"
+  val POOL_DEFAULT_SCHEDULING_MODE = SchedulingMode.FIFO
+  val POOL_DEFAULT_MINIMUM_SHARES = 2
+  val POOL_DEFAULT_WEIGHT = 1
```
For these, there's probably no need for the POOL_ in front. You can just call them DEFAULT_POOL_NAME, MINIMUM_SHARES_PROPERTY, etc.

mateiz commented on core/src/main/scala/spark/scheduler/cluster/FairTaskSetQueuesManager.scala:
```diff
@@ -0,0 +1,157 @@
+
+        val xmlSchedulingMode = (poolNode \ POOL_SCHEDULING_MODE_PROPERTY).text
+        if (xmlSchedulingMode != "") {
+          try{
```
When opening blocks like this, leave a space before the "{"; so for example:

```
try {
```

mateiz commented on core/src/main/scala/spark/scheduler/cluster/Pool.scala:
```diff
@@ -0,0 +1,74 @@
+import scala.collection.mutable.ArrayBuffer
+
+import spark.Logging
+import spark.scheduler.cluster.SchedulingMode.SchedulingMode
```
Add a space after the imports

mateiz commented on core/src/main/scala/spark/scheduler/cluster/FairTaskSetQueuesManager.scala:
```diff
@@ -0,0 +1,157 @@
+  val poolScheduleAlgorithm = new FairSchedulingAlgorithm()
+  val POOL_FAIR_SCHEDULER_PROPERTIES = "spark.scheduler.cluster.fair.pool"
+  val POOL_DEFAULT_POOL_NAME = "default"
+  val POOL_MINIMUM_SHARES_PROPERTY = "minShares"
```
I would just call this minShare

mateiz commented on core/src/main/scala/spark/scheduler/cluster/Pool.scala:
```diff
@@ -0,0 +1,74 @@
+      schedulingMode match {
+      case SchedulingMode.FAIR =>
+        val schedule = new FairSchedulingAlgorithm()
+        schedule
```
Here, there's no need to create a val; you can just do

```
case SchedulingMode.FAIR =>
  new FairSchedulingAlgorithm
```

mateiz commented on core/src/main/scala/spark/scheduler/cluster/Pool.scala:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  def addTaskSetManager(manager:TaskSetManager) {
```
Leave a space after the ":"

mateiz commented on core/src/main/scala/spark/scheduler/cluster/Pool.scala:
```diff
@@ -0,0 +1,74 @@
+    val sortedActiveTasksSetQueue = activeTaskSetsQueue.sortWith(taskSetSchedulingAlgorithm.comparator)
+    for (manager <- sortedActiveTasksSetQueue) {
+    logDebug("poolname:%s,taskSetId:%s,taskNum:%d,minShares:%d,weight:%d,runningTasks:%d".format(
+      poolName,manager.taskSet.id,manager.numTasks,manager.minShare,manager.weight,manager.runningTasks))
```
Indentation here looks wrong

mateiz commented on core/src/main/scala/spark/scheduler/cluster/Pool.scala:
```diff
@@ -0,0 +1,74 @@
+        if (task != None) {
+          manager.runningTasks += 1
+          return task
+        }
```
Too many spaces of indentation within the loop (looks like 4 spaces, or maybe a tab)

mateiz commented on core/src/main/scala/spark/scheduler/cluster/SchedulingAlgorithm.scala:
```diff
@@ -0,0 +1,55 @@
+        val stageId2 = s2.stageId
+        res = Math.signum(stageId1 - stageId2)
+    }
+    if (res < 0)
```
Put curly braces around the if and else bodies

mateiz commented on core/src/main/scala/spark/scheduler/cluster/SchedulingAlgorithm.scala:
```diff
@@ -0,0 +1,55 @@
+    else if (s1Needy && s2Needy)
+      res = minShareRatio1 <= minShareRatio2
+    else
+      res = taskToWeightRatio1 <= taskToWeightRatio2
```
Put curly braces around the if and else bodies since this is a pretty long statement

mateiz commented on core/src/main/scala/spark/scheduler/cluster/TaskSetManager.scala:
```diff
@@ -35,7 +41,12 @@ private[spark] class TaskSetManager(sched: ClusterScheduler, val taskSet: TaskSe
   val ser = SparkEnv.get.closureSerializer.newInstance()
 
+  var weight = TASKSET_WEIGHT
+  var minShare = TASKSET_MINIMUM_SHARES
```
You can probably just initialize these to 1 and 0 instead of having variables called TASKSET_WEIGHT and TASKSET_MINIMUM_SHARES. Also, I believe a default minShare of 0 works, right? In the Hadoop fair scheduler, the minShare only had to be set for pools that need special service.

mateiz commented on core/src/main/scala/spark/scheduler/cluster/TaskSetQueuesManager.scala:
```diff
@@ -0,0 +1,16 @@
+ * An interface for managing TaskSet queue/s that allows plugging different policy for
+ * offering tasks to resources
+ */
+private[spark] trait TaskSetQueuesManager {
```
I would rename this to something like QueueManager, since it's a pretty long name; and then, rename the subclasses to FairQueueManager and FIFOQueueManager.

mateiz commented on core/src/main/scala/spark/scheduler/cluster/TaskSetManager.scala:
```diff
@@ -222,7 +233,7 @@ private[spark] class TaskSetManager(sched: ClusterScheduler, val taskSet: TaskSe
             taskSet.id, index, serializedTask.limit, timeTaken))
           val taskName = "task %s:%d".format(taskSet.id, index)
-          return Some(new TaskDescription(taskId, execId, taskName, serializedTask))
+          return Some(new TaskDescription(taskId,taskSet.id,execId, taskName, serializedTask))
```
Put spaces after commas

mateiz commented on core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
```diff
@@ -122,20 +124,26 @@ private[spark] class ClusterScheduler(val sc: SparkContext)
    */
   def resourceOffers(offers: Seq[WorkerOffer]): Seq[Seq[TaskDescription]] = {
-    synchronized {
+     synchronized {
```
Seems like an extra space snuck in here

mateiz commented on core/src/main/scala/spark/SparkContext.scala:
```diff
@@ -139,7 +146,9 @@ class SparkContext(
         val backend = new SparkDeploySchedulerBackend(scheduler, this, sparkUrl, appName)
-        scheduler.initialize(backend)
+        val taskSetQueuesManager = Class.forName(System.getProperty("spark.cluster.taskscheduler")).
+                                   newInstance().asInstanceOf[TaskSetQueuesManager]
```
Here, the rest of the code puts dots on the line below, like this:

```
Class.forName(System.getProperty(...))
     .newInstance().asInstanceOf[...]
```

mateiz commented on core/src/main/scala/spark/scheduler/cluster/FairTaskSetQueuesManager.scala:
```diff
@@ -0,0 +1,157 @@
+         return task
+     }
+   }
+   return None
```
Indentation is weird here (looks like only 1 space)

mateiz commented:
Hey Andrew,

Thanks for the updates so far, and sorry for taking a bit of time to look at this in more detail -- I've been traveling a bunch these past few weeks. I took a more thorough look through the pull request and it seems pretty good overall, but I've pointed out some minor issues, mostly formatting fixes. (Not sure whether you're done with those since I saw you committed some more fixes today.) I also had a few bigger suggestions:

1) Rename the minShares XML property to minShare. That will match the Hadoop scheduler a bit better.

2) Rename the class TaskSetQueueManager to QueueManager, and similarly for its subclasses. It's a little long.

3) Would it be possible to make receiveOffer a part of the Schedulable interface? That would allow us to have nested pools in the future, and would just make the design a little cleaner. You might have to make other things part of the interface too, like taskFinished and checkSpeculativeTasks.

4) It's unfortunate that this scheduler doesn't work in local mode. I guess one solution would be to make local mode a subclass of ClusterScheduler too eventually. I'm fine doing that later, but if you think you could set up pools in the same way in both local and cluster schedulers, that would be nice.

5) Right now there are no tests for this -- it would be good to add some unit tests for the scheduling order (at least for SchedulingAlgorithm, but hopefully one end-to-end test that checks that we are launching tasks fairly between multiple jobs).

xiajunluan committed:
Contiue to update codes for scala code style
1.refactor braces for "class" "if" "while" "for" "match"
2.make code lines less than 100
3.refactor class parameter and extends defination

xiajunluan commented:
Hi Mateiz

I confuse why we will support nested pools in the futures, Does it mean that pool will contain other pools? if so, I think that I should merge the interface Schedulable and TasksetQueuesManager for a new Schedulable interface and discard the class FIFOTaskSetQueuesManager and FairTaskSetQueuesManager, and finally diagram will look like
![Fair Scheduler two version](https://f.cloud.github.com/assets/1168928/356966/cd312a18-a11b-11e2-8d81-85649bd70f66.jpg)

BTW, I will breakdown this feature for two steps
1.In this pull request, I will fix the format issues, refactor the new interface and construct unit test for SchedulingAlgorithm
2. set up pools for local and cluster together and implement end-to-end unit test in another pull request.

Any suggestion?

mateiz commented:
Hi Andrew,

That sounds good in terms of a plan. And yes, with a more general Schedulable interface, it means we could have pools contain other pools, which I think would be useful. The design you drew for it looks good.

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

xiajunluan committed:
refactor the Schedulable interface and add unit test for SchedulingAlgorithm

xiajunluan commented:
Hi Matei

I have refactored the schedulable interface to support nested pools, and also added SchedulableBudiler interface to build total Schedulable tree for FIFO and Fair, finally construct the unit test for schedulingAlgorithm.

xiajunluan committed:
remove TaskSetQueueManager and update code style

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

mateiz commented:
Hey Andrew, just an update on my side -- we're currently debugging some issues caused by an earlier merge (https://github.com/mesos/spark/pull/575), but once that's done, I'm going to add this pull request in as well (I can manually handle the update to work with Mridul's code). So it hasn't been forgotten! I hope to get to it very soon.

xiajunluan commented:
Hi Matei

I will rebase this feature to latest master branch and avoid you to resolve conflicts manually, Does it make sense for you?

mateiz commented:
Please do it if you have time. Otherwise I can do it myself. A few things changed due to the YARN patch adding rack locality.

Matei

On May 15, 2013, at 10:39 PM, Andrew xia notifications@github.com wrote:

> Hi Matei
> 
> I will rebase this feature to latest master branch and avoid you to resolve conflicts manually, Does it make sense for you?
> 
> —
> Reply to this email directly or view it on GitHub.

xiajunluan committed:
Fix ClusterScheduler bug to avoid allocating tasks to same slave

xiajunluan committed:
expose TaskSetManager type to resourceOffer function in ClusterScheduler

xiajunluan committed:
Merge branch 'master' into xiajunluan

Conflicts:
	core/src/main/scala/spark/SparkContext.scala
	core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala
	core/src/main/scala/spark/scheduler/cluster/TaskSetManager.scala

xiajunluan commented:
Hi Matei
I have merged this feature with latest master branch, and it has passed all unit tests. I also have tested fair scheduler algorithm in our internal cluster for several days.

xiajunluan committed:
fix bug of unit tests

mateiz commented:
Jenkins: this is ok to test

mateiz commented:
Thanks Andrew. I've run the tests on it and merged this in.
